### PR TITLE
Add dynamic session cache which allocates sessions from the heap.

### PR DESCRIPTION
**Pull request by kareem-wolfssl**
# Description

Adds a dynamically allocated session cache.

Fixes zd#14646

# Testing

Built in tests including API test.

# Checklist

 - [ ] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

### CODE REVIEW

File: src/ssl.c [hunk 1]
```diff
@@ -6190,7 +6190,11 @@ int AddCA(WOLFSSL_CERT_MANAGER* cm, DerBuffer** pDer, int type, int verify)
     typedef struct SessionRow {
         int nextIdx;                           /* where to place next one   */
         int totalCount;                        /* sessions ever on this row */
+#ifdef SESSION_CACHE_DYNAMIC_MEM
```
Comment (dgarske) [inline src/ssl.c:6193]: Please document the new `SESSION_CACHE_DYNAMIC_MEM` option at top of ssl.c.

File: src/ssl.c [hunk 2]
```diff
@@ -11469,7 +11473,13 @@ int wolfSSL_SetServerID(WOLFSSL* ssl, const byte* id, int len, int newSession)
 
 #endif /* !NO_CLIENT_CACHE */
 
-#if defined(PERSIST_SESSION_CACHE)
+/* TODO: Add SESSION_CACHE_DYNAMIC_MEM support for PERSIST_SESSION_CACHE.
```
Comment (dgarske) [inline src/ssl.c:11476]: Consider adding a settings.h rule to not allow `SESSION_CACHE_DYNAMIC_MEM` with `PERSIST_SESSION_CACHE`?
Please add `*` for comment section.

File: src/ssl.c [hunk 3]
```diff
@@ -14223,6 +14236,16 @@ int wolfSSL_Cleanup(void)
     }
     session_lock_valid = 0;
     #endif
+    #ifdef SESSION_CACHE_DYNAMIC_MEM
+    for (i = 0; i < SESSION_ROWS; i++) {
+        for (j = 0; j < SESSIONS_PER_ROW; j++) {
+            if (SessionCache[i].Sessions[j]) {
+                XFREE(&SessionCache[i].Sessions[j], NULL, DYNAMIC_TYPE_SESSION);
```
Comment (dgarske) [inline src/ssl.c:14243]: Does the `&SessionCache[i].Sessions[j]` need the leading `&`? Is heap hint not available here?

File: src/ssl.c [hunk 4]
```diff
@@ -15265,7 +15305,21 @@ int AddSessionToCache(WOLFSSL_CTX* ctx, WOLFSSL_SESSION* addSession,
...
         *sessionIndex = (row << SESSIDX_ROW_SHIFT) | idx;
 #endif
+
+#ifdef SESSION_CACHE_DYNAMIC_MEM
+    cacheSession = sessRow->Sessions[idx];
+    if (cacheSession) {
+        XFREE(cacheSession, NULL, DYNAMIC_TYPE_SESSION);
+        cacheSession = NULL;
+    }
+    cacheSession = (WOLFSSL_SESSION*) XMALLOC(sizeof(WOLFSSL_SESSION), NULL,
```
Comment (dgarske) [inline src/ssl.c:15315]: Same, is heap hint available?

File: src/ssl.c [hunk 5]
```diff
@@ -15697,8 +15762,14 @@ static int get_locked_session_stats(word32* active, word32* total, word32* peak)
 
         for (; count > 0; --count) {
             /* if not expired then good */
+#ifdef SESSION_CACHE_DYNAMIC_MEM
+            if (row->Sessions[idx] &&
+                ticks < (row->Sessions[idx]->bornOn +
+                            row->Sessions[idx]->timeout) ) {
```
Comment (dgarske) [inline src/ssl.c:15768]: Move the `{` out of the macro logic.
```
{
    now++;
}
```

File: src/ssl.c [hunk 6]
```diff
@@ -34242,6 +34325,11 @@ int wolfSSL_SSL_CTX_remove_session(WOLFSSL_CTX *ctx, WOLFSSL_SESSION *s)
                     cacheSession->ownExData = 0; /* We clear below */
                     s->ownExData = 1;
 #endif
+
+#ifdef SESSION_CACHE_DYNAMIC_MEM
+                    XFREE(cacheSession, NULL, DYNAMIC_TYPE_SESSION);
```
Comment (dgarske) [inline src/ssl.c:34330]: Heap hint?

**Review by dgarske — CHANGES_REQUESTED** [review]

**Comment by kareem-wolfssl** [comment]
Heap hint added, but I had to add it to SessionRow, so SessionCache will have SESSION_ROWS heap hints/pointers.
Not sure if there's a better spot for it, I need a heap hint which is available in wolfSSL_Init/Cleanup.

**Comment by dgarske** [comment]
Retest this please

**Review by dgarske — CHANGES_REQUESTED** [review]
Kareem, is it possible to have `make check` work with this enabled?

```
./configure --enable-all --disable-savesession CFLAGS="-DSESSION_CACHE_DYNAMIC_MEM" && make

$ ./tests/unit.test 
starting unit tests...
 Begin API Tests
     1: test_fileAccess                                     : skipped
     2: test_wolfSSL_Init                                   : passed (  0.00015)
     3: test_wolfSSL_Method_Allocators                      : passed (  0.00005)
     4: test_wolfSSL_CTX_new                                : passed (  0.00007)
     5: test_for_double_Free                                : passed (  0.00145)
     6: test_wolfSSL_get_finished                           :Client message: hello wolfssl!
Server response: I hear you fa shizzle!
 passed (  0.02621)
     7: test_wolfSSL_CTX_add_session                        :Client message: hello wolfssl!
Server response: I hear you fa shizzle!
Client message: hello wolfssl!

ERROR - tests/api.c line 7259 failed with:
Server response: I hear you fa shizzle!
    expected: wolfSSL_session_reused(ssl) == 1
    result:   0 != 1
ERROR - tests/api.c line 7259 failed with:
    expected: wolfSSL_session_reused(ssl) == 1
    result:   0 != 1



Aborted (core dumped)
```

**Review by dgarske — APPROVED** [review]

**Review by SparkiDev — APPROVED** [review]

**Pull request merged**